src/main.ts

```diff
@@ -0,0 +1,215 @@
+import { Euler, Quaternion, Vector3 } from 'three'
+import { R4AA } from './r4aa'
+
+const angleEpsilon = 0.0001
+
+const xAxis = new Quaternion(-1, 0, 0, 0)
+const zAxis = new Quaternion(0, 0, +1, 0)
+const quatA = new Quaternion()
+const quatB = new Quaternion()
+const quatC = new Quaternion()
+const quatD = new Quaternion()
+const quatE = new Quaternion()
+
+const vecA = new Vector3()
+const vecB = new Vector3()
+const vecC = new Vector3()
+const vecD = new Vector3()
+
+const r4aa = new R4AA()
+
+const defaultAngleEpsilon = 0.000_01
+
+const anglesToQuat = (dest: Quaternion, angle1: number, angle2: number, angle3: number): Quaternion => {
+  const y = [0, 0, 0]
+  const x = [0, 0, 0]
+
+  y[0] = Math.sin(angle1 / 2)
+  x[0] = Math.cos(angle1 / 2)
+
+  y[1] = Math.sin(angle2 / 2)
+  x[1] = Math.cos(angle2 / 2)
+
+  y[2] = Math.sin(angle3 / 2)
+  x[2] = Math.cos(angle3 / 2)
+
+  return dest.set(
+    (x[0] * y[1] * y[2]) - (y[0] * y[1] * x[2]),
+    (x[0] * y[1] * x[2]) + (y[0] * y[1] * y[2]),
+    (y[0] * x[1] * x[2]) + (x[0] * x[1] * y[2]),
+    (x[0] * x[1] * x[2]) - (y[0] * x[1] * y[2])
+  )
+}
+
+// See: https://github.com/viamrobotics/rdk/blob/main/spatialmath/orientationVector.go
+/**
+ * Viam’s orientation vector is a method for describing the orientation of an object in 3D space.
+ * It is part of a Pose which also includes the position in 3D space.
+ * 
+ * The vector extends from the center of the object to another point in the reference frame. This defines the direction something is pointing in.
+ */
+export class OrientationVector {
+  readonly isOrientationVector = true
+
+  /**
+   * Describes the angular position around the vector.
+   */
+  th: number
+
+  /**
+   * The vector's x component.
+   */
+  x: number
+
+  /**
+   * The vector's y component.
+   */
+  y: number
+
+  /**
+   * The vector's z component.
+   */
+  z: number
+
+  constructor (th = 0, x = 0, y = 0, z = 0) {
+    this.th = th
+    this.x = x
+    this.y = y
+    this.z = z
+  }
+
+  length (): number {
+    return Math.sqrt((this.x ** 2) + (this.y ** 2) + (this.z ** 2))
+  }
+
+  normalize (): this {
+    const length = this.length()
```

Comment:
Is it worth handling the zero vector case here?

Reply:
Yeah, good catch, I've just hijacked Three.js' `normalize()` which does just that.